xla/service/gpu/gpu_all_gather_optimizer.h

```diff
@@ -0,0 +1,42 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef XLA_SERVICE_GPU_GPU_ALL_GATHER_OPTIMIZER_CREATOR_H_
+#define XLA_SERVICE_GPU_GPU_ALL_GATHER_OPTIMIZER_CREATOR_H_
+
+#include "tensorflow/compiler/xla/service/hlo_pass_interface.h"
+
+namespace xla {
+namespace gpu {
+
+// Transforms binary_op(all-gather(reduce_scatter(a)),
+// all-gather(reduce_scatter(b))) to allgather(binary_op(reduce_scatter(a),
+// reduce_scatter(b)))
+
+class AllGatherOptimizer : public HloModulePass {
```

Comment:
You could make this a subclass of OpExpanderPass, I think it would make the code a little bit better structured.
The InstructionMatchesPattern function can then check for

```
      return instruction->opcode() == HloOpcode::kReduceScatter &&
          !instruction->shape().IsArray();
```

And you will not need to loop through the non-fusion computations yourself.

The ExpandInstruction function can then just return a nullptr if MatchReduceScatter returns std::nullopt.

Reply:
I added the OpExpanderPass and changed the code accordingly. While testing the code, faced the following error:
`Operand 0 (reduce-scatter.1) of instruction add.2 is in a different computation: (null) vs main`

for the following example hlo:

`
param.1 = bf16[8,128,1024]{2,1,0} parameter(0)
param.2 = bf16[8,128,1024]{2,1,0} parameter(1)
reduce-scatter.1 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
all-gather.1 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
reduce-scatter.2 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.2), channel_id=9, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
all-gather.2 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.2), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
add.1 = bf16[8,128,1024]{2,1,0} add(all-gather.1, all-gather.2)
`
While checking the OpExpanderPass class code, I noticed that in the MatchReduceScatter phase the operations are cached and within the ExpandInstruction they are processed as you mentioned. Now, each instruction is assigned a computation at this point but my code might modify one while checking the other (as in the above example) resulting the issue. I suggest keeping the class inherit from `HloModulePass` unless I am missing something, in which case, please kindly point me to the right direction.